design-proposals/klip-5-internal-schemas-for-KSQL-engine.md

```diff
@@ -0,0 +1,191 @@
+based on the internal schema. It then will use the rewritten query to process the messages from the topic.
+
+## Scope
+
+* Support for Quoted Identifiers in DDL and DML statements
+* Creation of internal schema for external schemas
+* Query rewrite to use the internal schemas
+* Query processing using the internal schemas
+
+## Value/Return
+
+This will eliminate the current requirement of having message field names in Kafka topics to conform
+to KSQL internal requirements. With this KLIP, KSQL will be able to process topics with any naming convention.
+
+
+## Public API
+
+The only user facing part in this KLIP is the support for quoted identifier which will be the same
+as the standard quoted identifiers in SQL. The declaration and use of the internal schemas are not
+visible to users and are part of the KSQL engine internals.
+
+## Design
+
+The internal schema will only be used in executing DML statements. When the engine receives a DML
+statement (CSAS, CTAS, INSERT INTO, SELECT), it builds an internal schema with valid field names for
+the given source schema(s). This will be done by extending the current KSQLSchema class by adding
+the internal schema to it. The internal schema field names can follow a predefined protocol so that
+KSQL engine would build the same internal schema for a given external schema everytime. As an example,
+ let’s consider we use the index of the field in the schema as a prefix to a constant string value to
+ build the internal schema field names. Consider the previous example where our external schema has
+ three fields as the following:
+
+```sql
+(“@ID” BIGINT, “@NAME” STRING, “MESSAGE.AMOUNT” DOUBLE)
+```
+
+Assuming the constant string prefix to build the internal schema fields is “COL_”, the internal
+schema for the above schema will be as the following:
+
+```sql
+(COL_0 BIGINT, COL_1 STRING, COL_2 DOUBLE)
+```
+
+Note that the following one to one mapping will exist between the fields from the external schema and the fields from the internal schema:
+
+* “@ID”                            <==>   COL_0
+* “@NAME”                     <==>   COL_1
+* “MESSAGE.AMOUNT” <==>   COL_2
+
+In the case of nested fields, we can apply the same approach to the nested fields inside ARRAY, MAP
+and STRUCT types.
+
+Now that we have the internal schema built, the engine will rewrite the queries by replacing the
+identifiers that refer to the fields in the external schema with their corresponding field name in
+the internal schema.
+
+As an example, consider the following query again:
+
+```sql
+CREATE STREAM bar AS
+SELECT “@ID”, lowercase(“@NAME”) AS “@NAME”, “MESSAGE.AMOUNT” * 100 AS “AMOUNT.BY.100”
+FROM foo
+WHERE “MESSAGE.AMOUNT” < 1;
+```
+
+With the above mapping the engine will write this query as the following:
+
+```sql
+
+CREATE STREAM bar AS
+SELECT FOO_COL_1 AS “@ID”, lowercase(FOO_COL_1) AS “@NAME”, FOO_COL_2 * 100 AS “AMOUNT.BY.100”
```

Comment:
Yes, but now the prefixing should be done as part of the rewrite in this section since we will have ambiguity if we have more than one source, for instance in JOIN queries.

Reply:
true dat.